apollo-router/src/uplink/entitlement_stream.rs

```diff
@@ -109,67 +113,88 @@ pub(crate) enum Error {
 pin_project! {
     /// This stream wrapper will cause check the current entitlement at the point of warn_at or halt_at.
     /// This means that the state machine can be kept clean, and not have to deal with setting it's own timers and also avoids lots of racy scenarios as entitlement checks are guaranteed to happen after an entitlement update even if they were in the past.
-    struct EntitlementExpander<Upstream>
+    pub(crate) struct EntitlementExpander<Upstream>
     where
-        Upstream: Stream<Item = Event>,
+        Upstream: Stream<Item = Entitlement>,
     {
         #[pin]
         checks: DelayQueue<Event>,
         #[pin]
-        upstream: Upstream,
+        upstream: Fuse<Upstream>,
     }
 }
 
 impl<Upstream> Stream for EntitlementExpander<Upstream>
 where
-    Upstream: Stream<Item = Event>,
+    Upstream: Stream<Item = Entitlement>,
 {
     type Item = Event;
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let mut this = self.project();
-        let active_checks = !this.checks.is_empty();
-        match this.checks.poll_expired(cx) {
+        let checks = this.checks.poll_expired(cx);
+        let upstream_terminated = this.upstream.is_terminated();
+        match checks {
             // We have an expired claim that needs checking
             Poll::Ready(Some(item)) => Poll::Ready(Some(item.into_inner())),
-
+            Poll::Ready(None) if upstream_terminated => Poll::Ready(None),
+            // No more upstream, Waiter scheduled in this.checks.poll_expired(cx);
```

Comment:
I'll be honest. I don't really follow the logic of the polling logic here. Missing the walkthough means it's very hard for me to follow this. I'm not saying there's anything wrong with it, just that I can't review it effectively.

I think introducing `fuse()` is a good idea...

Reply:
Significant refactoring occurred and we had a sync walkthrough to resolve this.